handlers/utils/compute_utils.go

```diff
@@ -0,0 +1,1186 @@
+				logrus.Error(fmt.Sprintf("container id %s fail to start", containerResponse.ID))
+				return "", err1
+			}
+			return containerResponse.ID, nil
+		}
+		return "", err
+	}
+	logrus.Info(containerResponse.ID)
+	return containerResponse.ID, nil
+}
+
+func removeContainer(client *client.Client, containerID string) error {
+	client.ContainerKill(context.Background(), containerID, "KILL")
+	err := client.ContainerRemove(context.Background(), containerID, types.ContainerRemoveOptions{})
+	if err == nil {
+		logrus.Infof("container ID [%v] removed", containerID)
+	}
+	return err
+}
+
+func DoInstancePull(params *model.ImageParams, progress *progress.Progress) (types.ImageInspect, error) {
+	client := docker.GetClient(DefaultVersion)
+
+	imageJSON, ok := GetFieldsIfExist(params.Image.Data, "dockerImage")
+	if !ok {
+		return types.ImageInspect{}, errors.New("field not exist")
+	}
+	var dockerImage model.DockerImage
+	mapstructure.Decode(imageJSON, &dockerImage)
+	// imageName := fmt.Sprintf("%s:%s", dockerImage.FullName)
+	existing, _, err := client.ImageInspectWithRaw(context.Background(), dockerImage.FullName, false)
+	if err != nil {
+		logrus.Error(err)
+	}
+	if params.Mode == "cached" && err == nil {
+		return existing, nil
+	}
+	if params.Complete {
+		logrus.Info("complete")
+		_, err1 := client.ImageRemove(context.Background(), dockerImage.FullName+params.Tag, types.ImageRemoveOptions{Force: true})
+		if err1 != nil {
+			logrus.Error(err1)
+		}
+		return types.ImageInspect{}, nil
+	}
+	pullImage(&params.Image, progress)
+
+	if len(params.Tag) > 0 {
+		imageInfo := parseRepoTag(dockerImage.FullName)
+		repoTag := fmt.Sprintf("%s:%s", imageInfo["repo"], imageInfo["tag"]+params.Tag)
+		logrus.Info(repoTag)
+		client.ImageTag(context.Background(), dockerImage.FullName, repoTag)
+	}
+
+	inspect, _, err2 := client.ImageInspectWithRaw(context.Background(), dockerImage.FullName, false)
+	logrus.Infof("image inspect %v", inspect)
+	return inspect, err2
+}
+
+func createContainerConfig(imageTag string, command strslice.StrSlice, createConfig map[string]interface{}) *container.Config {
+	createConfig["image"] = imageTag
+	var config container.Config
+	err := mapstructure.Decode(createConfig, &config)
+	res, _ := json.Marshal(config)
+	logrus.Infof("config created %v", string(res))
+	if err != nil {
+		logrus.Error(err)
+	}
+	return &config
+}
+
+func getImageTag(instance *model.Instance) (string, error) {
+	var dockerImage model.DockerImage
+	mapstructure.Decode(instance.Image.Data["dockerImage"], &dockerImage)
+	imageName := dockerImage.FullName
+	if imageName == "" {
+		return "", errors.New("Can not start container with no image")
+	}
+	return imageName, nil
+}
+
+func isTrue(instance *model.Instance, field string) bool {
+	value, ok := GetFieldsIfExist(instance.Data, "fields", field)
+	if ok {
+		return InterfaceToBool(value)
+	}
+	return false
+}
+
+func setupDNSSearch(startConfig map[string]interface{}, instance *model.Instance) {
```

Comment:
Getfieldifexist is just a utility function to get fields inside a map without writing ugly type assertion. Ok is just bool value.